workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceFileSystemService.java

```
@@ -159,6 +160,98 @@ public boolean moveFile(final int workspaceId, final Path oldFilePath, final Pat
+      }
+
+      return true;
+    } catch (IOException e) {
+      e.printStackTrace();
+      return false;
+    }
+  }
+
+
+  @Override
+  public boolean copyDirectory(final int sourceWorkspaceId, final Path sourceFilePath, final int destWorkspaceId, final Path destFilePath)
+  throws NoSuchWorkspaceException, WorkspaceFileOpException
+  {
+    final var sourceRepoPath = postgresRepository.workspaceRootPath(sourceWorkspaceId);
+    final var sourcePath = sourceRepoPath.resolve(sourceFilePath);
+    final var destRepoPath = postgresRepository.workspaceRootPath(destWorkspaceId);
+    final var destPath = destRepoPath.resolve(destFilePath);
+
+    try {
+      // Validate source exists and is a directory
+      if (!Files.exists(sourcePath)) throw new WorkspaceFileOpException("Source directory \"%s\" in workspace %d does not exist.".formatted(sourceFilePath, sourceWorkspaceId));
+      if (!Files.isDirectory(sourcePath)) throw new WorkspaceFileOpException("Source directory \"%s\" in workspace %d is not actually a directory. This is a bug.".formatted(sourceFilePath, sourceWorkspaceId));
+
+      // Do not copy if destination already exists
+      if (Files.exists(destPath)) throw new WorkspaceFileOpException("Destination directory \"%s\" in workspace %d already exists.".formatted(destFilePath, destWorkspaceId));
+
+
+      // Walk source directory and copy files/subdirectories -- note we have to use a try-with-resources thing here
+      // to ensure the stream autocloses
+      try (var paths = Files.walk(sourcePath)) {
+        paths.forEach(source -> {
+          final Path relative = sourcePath.relativize(source);
+          final Path target = destPath.resolve(relative);
+          try {
+            if (Files.isDirectory(source)) {
+              Files.createDirectories(target);
+            } else {
+              Files.copy(source, target);
+            }
+          } catch (IOException e) {
+            throw new UncheckedIOException(e);
+          }
```

If you're not going to handle the IOException, you should let it be thrown as opposed to rewrapping it as Unchecked (especially since you catch `IOException` further on in the method).

---

Annoyingly Java doesn't let you throw checked exceptions like `IOException` unless you declare them, which you can’t do with an inline lambda. So it seems like you could either extract this into a separate function, or throw something else like a `RuntimeException`, but that's even less descriptive. Are either of those much of an improvement/more idiomatic than just doing this? `¯\_(ツ)_/¯`

---

No, if the exception is being wrapped in `Unchecked` to get around lambda restrictions, then `UncheckedIOException` makes the most sense.